File: packages/schemas-python/src/prism_schemas/trace.py
"""Trading-R1 reasoning trace schema."""

from __future__ import annotations

import hashlib
import json
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field
from prism_schemas.verdict import EvidenceToolReceipt


class Evidence(BaseModel):
    """A single piece of evidence supporting the thesis."""

    source: str
    claim: str
    confidence: float = Field(ge=0.0, le=1.0)
    timestamp: datetime


class ThesisStep(BaseModel):
    """A single step in the thesis composition."""

    proposition: str
    supporting_evidence_ids: list[int]
    risk_factors: list[str]


class TradingR1Trace(BaseModel):
    """Complete Trading-R1 structured reasoning trace."""

    trace_id: str
    agent_id: int
    market_id: str
    market_question: str

    thesis: list[ThesisStep]
    evidence: list[Evidence]

    # Tool-sourced evidence receipts — proves which evidence came from
    # real tools (verifiable) vs which the LLM invented (unverifiable).
    # Defaults to empty list for backward compatibility with old traces.
    evidence_receipts: list[EvidenceToolReceipt] = Field(default_factory=list)

    raw_probability: float = Field(ge=0.0, le=1.0)
    volatility_adjustment: float
    final_probability: float = Field(ge=0.0, le=1.0)

    action: Literal["BUY", "SELL", "HOLD"]
    size_usdc: float
    price_limit: float
    rationale: str

    model_family: Literal["anthropic-claude", "openai-gpt"]
    model_name: str
    created_at: datetime

    def content_hash(self) -> bytes:
        """Deterministic hash for on-chain anchoring."""
        canonical = json.dumps(self.model_dump(mode="json"), sort_keys=True)
        return hashlib.sha256(canonical.encode()).digest()

File: packages/schemas-typescript/src/trace.ts
/**
 * Trading-R1 reasoning trace schema — Zod v4 mirror of Python Pydantic model.
 */

import { z } from "zod/v4";
import { EvidenceToolReceiptSchema } from "./verdict";

/** A single piece of evidence supporting the thesis. */
export const EvidenceSchema = z.object({
  source: z.string(),
  claim: z.string(),
  confidence: z.number().min(0).max(1),
  timestamp: z.string(),
});

export type Evidence = z.infer<typeof EvidenceSchema>;

/** A single step in the thesis composition. */
export const ThesisStepSchema = z.object({
  proposition: z.string(),
  supporting_evidence_ids: z.array(z.number()),
  risk_factors: z.array(z.string()),
});

export type ThesisStep = z.infer<typeof ThesisStepSchema>;

/** Complete Trading-R1 structured reasoning trace. */
export const TradingR1TraceSchema = z.object({
  trace_id: z.string(),
  agent_id: z.number(),
  market_id: z.string(),
  market_question: z.string(),

  thesis: z.array(ThesisStepSchema),
  evidence: z.array(EvidenceSchema),

  // Tool-sourced evidence receipts — proves which evidence came from
  // real tools (verifiable) vs which the LLM invented (unverifiable).
  // Defaults to empty list for backward compatibility with old traces.
  evidence_receipts: z.array(EvidenceToolReceiptSchema).default([]),

  raw_probability: z.number().min(0).max(1),
  volatility_adjustment: z.number(),
  final_probability: z.number().min(0).max(1),

  action: z.enum(["BUY", "SELL", "HOLD"]),
  size_usdc: z.number(),
  price_limit: z.number(),
  rationale: z.string(),

  model_family: z.enum(["anthropic-claude", "openai-gpt"]),
  model_name: z.string(),
  created_at: z.string(),
});

export type TradingR1Trace = z.infer<typeof TradingR1TraceSchema>;
